crates/ruma-state-res/src/event_auth.rs

```diff
@@ -700,28 +700,26 @@ fn valid_membership_change(
             {
                 warn!("Join rule is not set to knock or knock_restricted, knocking is not allowed");
                 false
-            } else {
+            } else if sender != target_user {
                 // 2. If `sender` does not match `state_key`, reject.
+                warn!(
+                    ?sender,
+                    ?target_user,
+                    "Can't make another user knock, sender did not match target"
+                );
+                false
+            } else {
                 // 3. If the `sender`'s current membership is not `ban`, `invite`, or `join`, allow.
                 // 4. Otherwise, reject.
-                if sender != target_user {
-                    warn!(
-                        ?sender,
-                        ?target_user,
-                        "Can't make another user join, sender did not match target"
-                    );
-                    false
-                } else if matches!(
-                    sender_membership,
-                    MembershipState::Ban | MembershipState::Invite | MembershipState::Join
-                ) {
-                    warn!(
-                        ?target_user_membership_event_id,
-                        "Membership state of ban, invite or join are invalid",
-                    );
-                    false
-                } else {
-                    true
+                match sender_membership {
+                    MembershipState::Ban | MembershipState::Invite | MembershipState::Join => {
+                        warn!(
+                            ?target_user_membership_event_id,
+                            "Membership state of ban, invite or join are invalid",
+                        );
+                        false
+                    }
+                    _ => true,
```

Comment:
By exhaustive match, I meant listing all of the possible membership states explicitly here ^^

Reply:
But it won't be exhaustive because the type is marked as non-exhaustive. And clippy will complain if we mark `MembershipState::Leave | _ => true`

Reply:
Oh, okay. Hm, I wonder if the match makes that much sense then. Maybe we should just have the `else { if ... }` => `else if` change and leave the rest alone?